lib/encoder/src/Netzer.cpp

```diff
@@ -0,0 +1,144 @@
+}
+
+bool Netzer::read(void) {
+  int num_bytes = m_spi.write(tx_buffer, WORDS, rx_buffer, WORDS);
+  // TODO: Add check for error bit?
+  if ((num_bytes == WORDS)) {
+    angularVelocityEstimation();
+    return true;
+  } else {
+    return false;
+  }
+}
+
+bool Netzer::readAsync(callback_ptr callback) {
+  std::scoped_lock<Mutex> lock(m_mutex);
+  m_callback = callback;
+  // return status
+  int status = m_spi.transfer(tx_buffer, WORDS, rx_buffer, WORDS, event_callback_t(this, &Netzer::spi_callback_debug));
+  // status = 0 => SPI transfer started
+  // status = -1 => SPI peripheral is busy
+  return (status == 0);
+}
+
+uint16_t Netzer::get_raw_data() {
+  return m_raw_data;
+}
+
+void Netzer::spi_callback_debug(int events) {
+  angularVelocityEstimation();
+  m_callback();
+}
+
+bool Netzer::reset(void) {
+  std::scoped_lock<Mutex> lock(m_mutex);
+
+  bool success = read();
+  m_offset_deg = rawAbsToDegrees(m_position_raw);
+
+  m_position_deg                 = 0;
+  m_angular_velocity_deg_per_sec = 0;
+
+  return success;
+}
+
+bool Netzer::getAngleDeg(float &angle) {
+  std::scoped_lock<Mutex> lock(m_mutex);
+  bool success = read();
+  angle        = m_position_deg;
+  return success;
+}
+
+bool Netzer::getAngularVelocityDegPerSec(float &speed) {
+  std::scoped_lock<Mutex> lock(m_mutex);
+  bool success = read();
+  speed        = m_angular_velocity_deg_per_sec;
+  return success;
+}
+
+float Netzer::getAngleDegAsync(void) {
+  return m_position_deg;
+}
+
+float Netzer::getAngularVelocityDegPerSecAsync(void) {
+  return m_angular_velocity_deg_per_sec;
+}
+
+float Netzer::rawAbsToDegrees(uint16_t raw) {
+  return raw / 65536.0 * 360.0;  // 16 bit resolution, 2^16 = 65536
+}
+
+// Wrapper function  (taken from AEAT6012 estimator, maybe we can
+// refactor both to use a general estimator function like this?)
+void Netzer::angularVelocityEstimation(void) {
+  // Wrap-around catch for angular velocity estimation:
+  //
+  // If prev_pos = 359 deg and cur_pos = 1 deg, then the change in position would be 1 - 359 = -358
+  //  => This is wrong, since the true change in position is 2
+  //
+  // To prevent this issue, we assume that the sampling rate is high enough and take the smaller of the following
+  // two expressions to be the absolute change in position
+  //   1. abs(cur_pos - prev_pos)
+  //   2. 360 - abs(cur_pos - prev_pos)
+  //
+  // The direction depends on whether we are considering case 1 or case 2:
+  //   1. Direction is positive if cur_pos > prev_pos, negative if cur_pos < prev_pos
+  //   2. Direction is positive if cur_pos < prev_pos, negative if cur_pos > prev_pos
+
+  // Not sure if I'm handling the payload correctly
+  uint32_t raw_data = ((static_cast<uint32_t>(rx_buffer[0]) << 24) | (static_cast<uint32_t>(rx_buffer[1]) << 16)) |
+                      ((static_cast<uint32_t>(rx_buffer[0]) << 8) | (static_cast<uint32_t>(rx_buffer[0])));
+
+  m_raw_data = (uint16_t)((raw_data << 8) >> 8);
```

Comment:
Sure, so the idea is that according to the datasheet the positional data is from bits 8 to 24 (16 bits) so I just pulled only those bits into the raw_data. Naming isn't the greatest for sure lol, do you think ```m_raw_positonal_data``` would be better?

Reply:
Raw data is fine, although bits [8:24] is 17 bits, not 16. I'm still asking around for which exact Netzer model we have, from the datasheet it seems like there are 17, 18, and 19-bit variants.
![image](https://user-images.githubusercontent.com/36703188/113523597-7960d200-9576-11eb-896f-a8b7add8cc39.png)
 
![image](https://user-images.githubusercontent.com/36703188/113523629-c47ae500-9576-11eb-82e4-67b881e27e8f.png)

 I guess for now just treat it as 17-bit data, I'll let you know when I figure out which one it is

Reply:
Also note that your current bit shifting wouldn't extract the bits that you want, because you're left shifting by 8 then right shifting by 8, so the result would be that the left-most 8 bits become 0 while the rest of the bits still end up where they started.

Reply:
Oh gotcha thanks. Also i'm guessing that the table in the datasheet that breakdown the bits and their meaning is referring to the F, 17 bit version huh?

Reply:
Breaking down the raw data, would it look something like this:
rx_buffer[1] ... rx_buffer[0]
Bit 27 (Ack), Bit 26 (Start) ... Bit 0

or

rx_buffer[1] ... rx_buffer[0]
Bit 0, ... Bit 26 (Start), Bit 27 (Ack)